src/snapshot/component_clone.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{
+    GgrsComponentSnapshot, GgrsSnapshots, LoadWorld, LoadWorldSet, Rollback, RollbackFrameCount,
+    SaveWorld, SaveWorldSet,
+};
+use bevy::prelude::*;
+use std::marker::PhantomData;
+
+/// A [`Plugin`] which manages snapshots for a [`Component`] `C` using [`Clone`].
+pub struct GgrsComponentSnapshotClonePlugin<C>
+where
+    C: Component + Clone,
+{
+    _phantom: PhantomData<C>,
+}
+
+type Snapshots<C> = GgrsSnapshots<C, GgrsComponentSnapshot<C>>;
+
+impl<C> Default for GgrsComponentSnapshotClonePlugin<C>
+where
+    C: Component + Clone,
+{
+    fn default() -> Self {
+        Self {
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<C> GgrsComponentSnapshotClonePlugin<C>
+where
+    C: Component + Clone,
+{
+    pub fn save(
+        mut snapshots: ResMut<Snapshots<C>>,
+        frame: Res<RollbackFrameCount>,
+        query: Query<(&Rollback, &C)>,
+    ) {
+        let components = query
+            .iter()
+            .map(|(&rollback, component)| (rollback, component.clone()));
+        let snapshot = GgrsComponentSnapshot::new(components);
+        snapshots.push(frame.0, snapshot);
+    }
+
+    pub fn load(
+        mut commands: Commands,
+        mut snapshots: ResMut<Snapshots<C>>,
+        frame: Res<RollbackFrameCount>,
+        mut query: Query<(Entity, &Rollback, Option<&mut C>)>,
+    ) {
+        let snapshot = snapshots.rollback(frame.0).get();
+
+        for (entity, rollback, component) in query.iter_mut() {
+            let snapshot = snapshot.get(rollback);
+
+            match (component, snapshot) {
+                (Some(mut component), Some(snapshot)) => *component = snapshot.clone(),
+                (Some(_), None) => {
+                    commands.entity(entity).remove::<C>();
+                }
+                (None, Some(snapshot)) => {
+                    commands.entity(entity).insert(snapshot.clone());
+                }
+                (None, None) => {}
+            }
+        }
+    }
+}
+
+impl<C> Plugin for GgrsComponentSnapshotClonePlugin<C>
+where
+    C: Component + Clone,
+{
+    fn build(&self, app: &mut App) {
+        app.init_resource::<Snapshots<C>>()
+            .add_systems(
+                SaveWorld,
+                (Snapshots::<C>::discard_old_snapshots, Self::save)
```

Comment:
I like how you only had to implement discard_old_snapshots once :)

Reply:
These are the little things I was really happy writing!